src/NuGet.Clients/NuGet.VisualStudio.Common/Telemetry/NuGetPowerShellUsageCollector.cs

```diff
@@ -0,0 +1,421 @@
+        public const string ReOpenAtStart = nameof(ReOpenAtStart);
+
+        // Const name for emitting when VS solution close or VS instance close.
+        public const string Name = nameof(Name);
+        public const string NuGetPowershellPrefix = "NuGetPowershellPrefix."; // Using prefix prevent accidental same name property collission from different type telemetry.
+        public const string NuGetVSSolutionClose = nameof(NuGetVSSolutionClose);
+        public const string NuGetVSInstanceClose = nameof(NuGetVSInstanceClose);
+        public const string SolutionCount = nameof(SolutionCount);
+        public const string PMCPowerShellLoadedSolutionCount = nameof(PMCPowerShellLoadedSolutionCount);
+        public const string PMUIPowerShellLoadedSolutionCount = nameof(PMUIPowerShellLoadedSolutionCount);
+
+        private int _solutionCount;
+        // _vsSolutionData hold telemetry data for current VS solution session.
+        private NuGetPowershellVSSolutionCloseEvent _vsSolutionData;
+        // _vsInstanceData hold telemetry for current VS instance session.
+        private readonly NuGetPowershellVSInstanceCloseEvent _vsInstanceData;
+        private TelemetryEvent _powerShellLoadEvent;
+        private object _lock = new object();
+
+        public NuGetPowerShellUsageCollector()
+        {
+            _vsSolutionData = new NuGetPowershellVSSolutionCloseEvent();
+            _vsInstanceData = new NuGetPowershellVSInstanceCloseEvent();
+
+            NuGetPowerShellUsage.PowerShellLoadEvent += NuGetPowerShellUsage_PMCLoadEventHandler;
+            NuGetPowerShellUsage.PowerShellCommandExecuteEvent += NuGetPowerShellUsage_PowerShellCommandExecuteEvent;
+            NuGetPowerShellUsage.InitPs1LoadEvent += NuGetPowerShellUsage_InitPs1LoadEvent;
+            NuGetPowerShellUsage.PMCWindowsEvent += NuGetPowerShellUsage_PMCWindowsEventHandler;
+            NuGetPowerShellUsage.SolutionOpenEvent += NuGetPowerShellUsage_SolutionOpenHandler;
+            NuGetPowerShellUsage.SolutionCloseEvent += NuGetPowerShellUsage_SolutionCloseHandler;
+            NuGetPowerShellUsage.VSInstanceCloseEvent += NuGetPowerShellUsage_VSInstanseCloseHandler;
+        }
+
+        private void NuGetPowerShellUsage_PMCLoadEventHandler(bool isPMC)
+        {
+            AddPowerShellLoadedData(isPMC, _vsSolutionData);
+        }
+
+        internal void AddPowerShellLoadedData(bool isPMC, NuGetPowershellVSSolutionCloseEvent vsSolutionData)
+        {
+            lock (_lock)
+            {
+                // PowerShellHost loaded first time, let's emit this to find out later how many VS instance crash after loading powershell.
+                if (!vsSolutionData._loadedFromPMC && !vsSolutionData._loadedFromPMUI)
+                {
+                    var telemetryEvent = new TelemetryEvent(NuGetPowerShellLoaded, new Dictionary<string, object>
+                        {
+                            { NuGetPowershellPrefix + LoadedFromPMC, isPMC }
```

Comment:
Rather than a boolean, would it be better to have a `vs.nuget.powershellloadreason` that has values like `pmc` or `pmui`?

Reply:
I wasn't suggesting replacing the `isPMC` variable/parameter everywhere, just the reported telemetry:

```cs
var telemetryEvent = new TelemetryEvent(NuGetPowerShellLoaded, new Dictionary<string, object>()
{
  { NuGetPowershellPrefix + "Trigger", isPMC ? "PMC" : "PMUI" }
};
```

Consider the point of view from someone querying the data. They see `vs.nuget.nugetpowershellprefix.loadedfrompmc == false`, but if they don't NuGet very well, they might wonder where did it load from then?

This made me notice, do you really want to use `NuGetPowershellPrefix` for this?  This has a different event name `vs.nuget.nugetpowershellloaded`, so duplicating `nugetpowershell` in the property name is redundant.